### DISCUSSION TITLE
Missing meshoptimizer dependency in @gltf-transform/functions

### DISCUSSION THREAD

**Discussion by siclait** [donmccurdy/glTF-Transform#735] [General]
I feel like I'm missing something obvious, but when I attempt to run my application that is using `@gltf-transform/functions@2.4.6`, I'm getting an error about a missing dependency:

```
node_modules/@gltf-transform/functions/dist/simplify.d.ts:2:15 - error TS2305: Module '"meshoptimizer"' has no exported member 'MeshoptSimplifier'.

2 import type { MeshoptSimplifier } from 'meshoptimizer';
```

My Typescript application has a transitive dependency on `meshoptimizer@0.16.1`, which is why I'm not seeing similar errors for other `meshoptimizer` imports. But MeshoptSimplifier was added between 0.16.1 and 0.18.1.

Why isn't `meshoptimizer` listed as a dependency in `@gltf-transform/functions` if it's being used by the package? How can I use the library without running into this error?

**Comment by donmccurdy** [comment]
Hi @siclait! I assume you're using `@gltf-transform/functions` directly and not through the CLI? And are you actually trying to use `simplify()` or is this error occurring when just doing unrelated things with the library?

 
The intended use of the `simplify` function would be:

https://gltf-transform.donmccurdy.com/functions.html#simplify

```javascript
import { simplify } from '@gltf-transform/functions';
import { MeshoptSimplifier } from 'meshoptimizer';

await document.transform(
  simplify({ simplifier: MeshoptSimplifier, ratio: 0.75 })
);
```

That is, you are supposed to install `meshoptimizer` and pass it into the `simplify` function, it is not a production dependency of the project.

But the [type-only import](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-3-8.html#type-only-imports-and-export) does reference meshoptimizer. I haven't seen the error before but I guess it kind of makes sense, if `meshoptimizer` types are used in an exported API then your TS application may try to resolve it. I wonder if removing references to meshoptimizer from exported APIs would fix this, and let `@gltf-transform/functions` keep it as a dev-only dependency...

Finally [skipLibCheck](https://www.typescriptlang.org/tsconfig#skipLibCheck) might be a workaround here.

> EDIT: Ok, it does look like imports from `meshoptimizer` just need to be omitted from exported APIs if it's going to remain a dev dependency, per https://stackoverflow.com/a/46011417/1314762. Feel free to report this as a bug if that's the issue you're seeing!

**Reply by siclait** [reply]
Thanks @donmccurdy! Yes, I'm using the library directly (not through the CLI) and I'm getting the error on importing `webp` from `@gltf-transform/functions`.

It looks like the issue I'm facing is the one in the StackOverflow post you referenced—the imports from `meshoptimizer` are a part of the exported API so I'm running into these TypeScript errors even though I'm not using anything from `meshoptimizer`. I'll open an issue.

**Reply by donmccurdy** [reply]
Copied to https://github.com/donmccurdy/glTF-Transform/issues/741, to be fixed in https://github.com/donmccurdy/glTF-Transform/pull/742. :)
